### PR TITLE
[AutoDiff] Support differentiation of loops.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Change predecessor enum generation to support loops.
  - Generate indirect predecessor enums for BBs in loops.
  - Handle boxed payloads of indirect enums.
- Traverse basic blocks in post-order post-dominance order.
  - This is necessary for computational correctness.
- Add loop tests (`for-in`, `while`, `repeat-while`, `break`/`continue`, nested).

---

Examples:
```swift
func for_loop(_ x: Float) -> Float {
  var result = x
  for _ in 1..<3 {
    result = result * x
  }
  return result
}
expectEqual((8, 12), valueWithGradient(at: 2, in: for_loop))
expectEqual((27, 27), valueWithGradient(at: 3, in: for_loop))
```
```swift
func nested_loop(_ x: Float, count: Int) -> Float {
  var outer = x
  for _ in 1..<count {
    outer = outer * x

    var inner = outer
    var i = 1
    while i < count {
      inner = inner + x
      i += 1
    }
    outer = inner
  }
  return outer
}
expectEqual((20, 22), valueWithGradient(at: 2, in: { x in nested_loop(x, count: 3) }))
expectEqual((104, 66), valueWithGradient(at: 4, in: { x in nested_loop(x, count: 3) }))
```

[Verified loop differentiation correctness](https://gist.github.com/dan-zheng/d153d950687820418dbae65517259ff0) against PyTorch and Tangent.

Exposed [TF-584](https://bugs.swift.org/browse/TF-584): derivative computation bug for repeat-while loops.
Exposed [TF-585](https://bugs.swift.org/browse/TF-585): AllocBoxToStack crash with `-O` for nested loop AD.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -3055,15 +3042,28 @@ class VJPEmitter final
...
-    return builder.createEnum(loc, pbStructVal, enumEltDecl, enumLoweredTy);
+    auto enumEltType = remapType(
+        enumLoweredTy.getEnumElementType(enumEltDecl, getModule()));
+    // If the enum element type does not have a box type (i.e. the enum case is
+    // not indirect), then directly create an enum.
+    auto boxType = dyn_cast<SILBoxType>(enumEltType.getASTType());
+    if (!boxType)
+      return builder.createEnum(loc, pbStructVal, enumEltDecl, enumLoweredTy);
+    // Otherwise, box the pullback struct value and create an enum.
+    auto *allocBox = builder.createAllocBox(loc, boxType);
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: This `alloc_box` is not being deallocated and may cause leaks - to be verified.
Loop AD leak checking tests are added in `test/AutoDiff/leakchecking.swift`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -3738,16 +3741,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
...
-
   /// Mapping from original basic blocks and original active values to
   /// corresponding adjoint block arguments.
   DenseMap<std::pair<SILBasicBlock *, SILValue>, SILArgument *>
       activeValueAdjointBBArgumentMap;
 
+  /// Mapping from original basic blocks to local adjoint values to be cleaned
+  /// up. This is populated when adjoint emission is run on one basic block and
+  /// cleaned before processing another basic block.
+  DenseMap<SILBasicBlock *, SmallVector<AdjointValue, 8>>
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Note: changed `blockLocalAdjointValues` to a `DenseMap` for more principled/predictable behavior.
Previously, using `SmallVector` caused bugs when the original entry BB isn't the last visited BB during adjoint generation.

File: test/AutoDiff/control_flow.swift [hunk 3]
```diff
@@ -519,4 +519,46 @@ ControlFlowTests.test("Enums") {
...
+    var result = x
+    for _ in 1..<3 {
+      result = result * x
+    }
+    return result
+  }
+  expectEqual((8, 12), valueWithGradient(at: 2, in: for_loop))
+  expectEqual((27, 27), valueWithGradient(at: 3, in: for_loop))
+
+  func while_loop(_ x: Float) -> Float {
```
Comment (rxwei) [inline test/AutoDiff/control_flow.swift]: Please also test a `repeat`-`while` loop.

**Review by rxwei — COMMENTED** [review]
~~I couldn't find any emission of `dealloc_box` and this is apparently leaking a lot more memory (all pullback structs in loops are kept alive). This needs to be fixed before merging.~~
Actually, this should be handled automatically when you `release_value` the pullback struct.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -1737,6 +1736,7 @@ static void dumpActivityInfo(SILFunction &fn,
     for (auto &inst : bb)
       for (auto res : inst.getResults())
         dumpActivityInfo(res, indices, activityInfo, s);
+    s << "\n";
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
    s << '\n';
```

**Review by rxwei — APPROVED** [review]
The patch is pretty clean. LGTM!

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Resolved - see [this comment](https://github.com/apple/swift/pull/25558#pullrequestreview-250897375). Deallocation is handled automatically during `release_value` for pullback structs.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done in https://github.com/apple/swift/pull/25558/commits/32f66294d96791f67f15234948b493b76b9d31da.

File: test/AutoDiff/control_flow.swift (hunk 3 above)
Comment (dan-zheng) [reply test/AutoDiff/control_flow.swift]: Added tests for `repeat-while` loops and `continue`/`break` statements in https://github.com/apple/swift/pull/25558/commits/32f66294d96791f67f15234948b493b76b9d31da.
Exposed [TF-584](https://bugs.swift.org/browse/TF-584): derivative computation bug for repeat-while loops. Thanks for pointing this out!

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please clean test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
Two tests fail only with optimizations, which I hadn't tested locally:
```
Failing Tests (2):
    Swift(linux-x86_64) :: AutoDiff/leakchecking.swift
    Swift(linux-x86_64) :: AutoDiff/control_flow.swift
```

```
bool swift::StackNesting::solve(): Assertion `!(BI.Block->getTerminator()->isFunctionExiting() && Bits.any()) && "stack location is missing dealloc"' failed.
```
Perhaps there are some missing deallocations in generated code. Will investigate later.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 5]
```diff
@@ -3055,15 +3042,31 @@ class VJPEmitter final
...
+    if (!boxType)
+      return builder.createEnum(loc, pbStructVal, enumEltDecl, enumLoweredTy);
+    // Otherwise, box the pullback struct value and create an enum.
+    auto *allocBox = builder.createAllocBox(loc, boxType);
+    auto *projectBox = builder.createProjectBox(loc, allocBox, /*index*/ 0);
+    builder.createStore(loc, pbStructVal, projectBox,
+                        getBufferSOQ(projectBox->getType().getASTType(), *vjp));
+    // NOTE(TF-585): `fix_lifetime` is generated to avoid AllocBoxToStack crash
+    // for nested loop AD.
+    // builder.createFixLifetime(loc, allocBox);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: And yet it's commented out :)

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks for the catch c:

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**